### PR TITLE
feat: metric fields can be parameterized by analysis arguments

### LINKED ISSUE #825: Make AnalysisTemplate more configurable
# Summary

There are a number of fields in AnalysisTemplate/ClusterAnalysisTemplate that are static (`consecutiveErrorLimit`, `interval`, `count`, etc). It would mean a lot of practical benefits to make them dynamically configurable.

# Use Cases
Having a single AnalysisTemplate you could define different threasholds for different applications (since they may have different tolerance to failures). The preferred experience IMO would be something like
```yaml
postPromotionAnalysis:
    templates:
    - templateName: error-rate
      failureLimit: 3
      interval: 5m
      count: 10
```

---
**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

### PR DESCRIPTION
**Pull request by khhirani**
Closes #825 

Signed-off-by: khhirani <kareena.hirani@gmail.com>

### CODE REVIEW

**Review by jessesuen — CHANGES_REQUESTED** [review]
I see you made changes to all of the types, but I don't see any tests which actually test the feature. Where is a test which actually does substitution for a metric like:

```yaml
  metrics:
  - name: error-rate
    provider:
      prometheus:
        interval: "{{args.interval}}"
        count: "{{args.count}}"
```

**Comment by jessesuen** [comment]
Also, an e2e would be nice, even if it is just modifying an existing one.

**Comment by jessesuen** [comment]
@khhirani you'll need to rebase on master to fix the codegen problem. I found a toolchain discrepancy between macOS kubectl vs. github's ubuntu-latest kubectl that causes the difference in manifests. I switched to using `kustomize build` instead of `kubectl kustomize`

**Comment by khhirani** [comment]
@jessesuen Fixed

File: docs/features/bluegreen.md [hunk 1]
```diff
@@ -57,14 +57,14 @@ spec:
...
       antiAffinity: object
       previewService: string
       prePromotionAnalysis: object
       postPromotionAnalysis: object
-      previewReplicaCount: *int32
-      scaleDownDelaySeconds: *int32
-      scaleDownDelayRevisionLimit: *int32
+      previewReplicaCount: intOrString
+      scaleDownDelaySeconds: intOrString
+      scaleDownDelayRevisionLimit: intOrString
```
Comment (jessesuen) [inline docs/features/bluegreen.md:67]: This is wrong -- you didn't change these types

File: utils/defaults/defaults.go [hunk 2]
```diff
@@ -111,8 +111,8 @@ func GetAutoPromotionEnabledOrDefault(rollout *v1alpha1.Rollout) bool {
 }
 
 func GetConsecutiveErrorLimitOrDefault(metric *v1alpha1.Metric) int32 {
-	if metric.ConsecutiveErrorLimit != nil {
-		return *metric.ConsecutiveErrorLimit
+	if &metric.ConsecutiveErrorLimit != nil && metric.ConsecutiveErrorLimit.IntValue() > 0 {
+		return int32(metric.ConsecutiveErrorLimit.IntValue())
```
Comment (jessesuen) [inline utils/defaults/defaults.go:115]: I think this is a bug. Doesn't this logic mean you are not allowing for `consecutiveErrorLimit: 0` ? Why was the check added?

File: rollout/analysis_test.go [hunk 3]
```diff
@@ -21,14 +21,16 @@ import (
...
 	return &v1alpha1.AnalysisTemplate{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      name,
 			Namespace: metav1.NamespaceDefault,
 		},
 		Spec: v1alpha1.AnalysisTemplateSpec{
 			Metrics: []v1alpha1.Metric{{
-				Name: "example",
+				Name:  "example",
+				Count: &count,
```
Comment (jessesuen) [inline rollout/analysis_test.go:33]: This seems to be changing the base template used by the tests. I would expect tests to behave differently because of this change because `count: nil` is different than `count: 0`.  Is this change necessary for some reason?

File: analysis/analysis_test.go [hunk 4]
```diff
@@ -1325,39 +1332,19 @@ func TestSecretNotFound(t *testing.T) {
...
-	tasks := []metricTask{{
-		metric: v1alpha1.Metric{
-			Name:             "metric-name",
-			SuccessCondition: "{{args.secret-empty}}",
-		},
-		incompleteMeasurement: nil,
-	}}
-	_, _, err := c.resolveArgs(tasks, args, metav1.NamespaceDefault)
-	assert.Equal(t, "arg 'secret-empty' does not contain a secret reference", err.Error())
-}
```
Comment (jessesuen) [inline analysis/analysis_test.go:1359]: What makes this test not valid anymore?

File: analysis/analysis_test.go [hunk 5]
```diff
@@ -1302,6 +1308,7 @@ func TestSecretContentReferenceAndMultipleArgResolutionSuccess(t *testing.T) {
 						}},
 					},
 				},
+				Count: &count,
```
Comment (jessesuen) [inline analysis/analysis_test.go:1311]: Why can't count be nil?

**Review by jessesuen — CHANGES_REQUESTED** [review]
I see that everywhere where count used to be nil, is now doing `count := intstr.FromInt(0)` and I'd like to understand why, since nil and zero mean different things.

File: utils/defaults/defaults.go (hunk 2 above)
Comment (khhirani) [reply utils/defaults/defaults.go]: That was my mistake, I had a pointer error in `analysis_types.go`. Fixed now.

File: analysis/analysis_test.go (hunk 4 above)
Comment (khhirani) [reply analysis/analysis_test.go:1359]: I modified the logic `analysis.go` so this error no longer occurs.

File: docs/features/bluegreen.md [hunk 6]
```diff
@@ -57,7 +57,7 @@ spec:
   strategy:
     blueGreen:
       autoPromotionEnabled: boolean
-      autoPromotionSeconds: *int32
+      autoPromotionSeconds: intOrString
```
Comment (jessesuen) [inline docs/features/bluegreen.md:60]: I don't think you changed this to intOrString

**Review by jessesuen — APPROVED** [review]
LGTM, please revert unintentional change to bluegreen.md before merging
